File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/services/ledger/ManageLedgerImpl.java
package com.decrux.pocketr.api.services.ledger;

import com.decrux.pocketr.api.entities.db.auth.User;
import com.decrux.pocketr.api.entities.db.ledger.Account;
import com.decrux.pocketr.api.entities.db.ledger.AccountType;
import com.decrux.pocketr.api.entities.db.ledger.CategoryTag;
import com.decrux.pocketr.api.entities.db.ledger.LedgerSplit;
import com.decrux.pocketr.api.entities.db.ledger.LedgerTxn;
import com.decrux.pocketr.api.entities.db.ledger.SplitSide;
import com.decrux.pocketr.api.entities.dtos.BalanceDto;
import com.decrux.pocketr.api.entities.dtos.CreateSplitDto;
import com.decrux.pocketr.api.entities.dtos.CreateTransactionDto;
import com.decrux.pocketr.api.entities.dtos.PagedTransactionsDto;
import com.decrux.pocketr.api.entities.dtos.SplitDto;
import com.decrux.pocketr.api.entities.dtos.TransactionDto;
import com.decrux.pocketr.api.entities.dtos.TxnCreatorDto;
import com.decrux.pocketr.api.exceptions.BadRequestException;
import com.decrux.pocketr.api.exceptions.ForbiddenException;
import com.decrux.pocketr.api.exceptions.NotFoundException;
import com.decrux.pocketr.api.repositories.AccountRepository;
import com.decrux.pocketr.api.repositories.CategoryTagRepository;
import com.decrux.pocketr.api.repositories.CurrencyRepository;
import com.decrux.pocketr.api.repositories.LedgerSplitRepository;
import com.decrux.pocketr.api.repositories.LedgerTxnRepository;
import com.decrux.pocketr.api.services.household.ManageHousehold;
import com.decrux.pocketr.api.services.ledger.validations.CrossUserAssetAccountTypeValidator;
import com.decrux.pocketr.api.services.ledger.validations.DoubleEntryBalanceValidator;
import com.decrux.pocketr.api.services.ledger.validations.HouseholdIdPresenceValidator;
import com.decrux.pocketr.api.services.ledger.validations.HouseholdMembershipValidator;
import com.decrux.pocketr.api.services.ledger.validations.HouseholdSharedAccountValidator;
import com.decrux.pocketr.api.services.ledger.validations.IndividualModeOwnershipValidator;
import com.decrux.pocketr.api.services.ledger.validations.MinimumSplitCountValidator;
import com.decrux.pocketr.api.services.ledger.validations.PositiveSplitAmountValidator;
import com.decrux.pocketr.api.services.ledger.validations.SplitSideValueValidator;
import com.decrux.pocketr.api.services.ledger.validations.TransactionAccountCurrencyValidator;
import com.decrux.pocketr.api.services.user_avatar.UserAvatarService;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class ManageLedgerImpl implements ManageLedger {

    private static final Set<AccountType> DEBIT_NORMAL_TYPES = Set.of(AccountType.ASSET, AccountType.EXPENSE);
    private static final Set<AccountType> TRANSFER_TYPES = Set.of(
        AccountType.ASSET,
        AccountType.LIABILITY,
        AccountType.EQUITY
    );

    private final LedgerTxnRepository ledgerTxnRepository;
    private final LedgerSplitRepository ledgerSplitRepository;
    private final AccountRepository accountRepository;
    private final CurrencyRepository currencyRepository;
    private final CategoryTagRepository categoryTagRepository;
    private final ManageHousehold manageHousehold;
    private final UserAvatarService userAvatarService;
    private final MinimumSplitCountValidator minimumSplitCountValidator;
    private final PositiveSplitAmountValidator positiveSplitAmountValidator;
    private final SplitSideValueValidator splitSideValueValidator;
    private final DoubleEntryBalanceValidator doubleEntryBalanceValidator;
    private final TransactionAccountCurrencyValidator transactionAccountCurrencyValidator;
    private final IndividualModeOwnershipValidator individualModeOwnershipValidator;
    private final HouseholdIdPresenceValidator householdIdPresenceValidator;
    private final HouseholdMembershipValidator householdMembershipValidator;
    private final HouseholdSharedAccountValidator householdSharedAccountValidator;
    private final CrossUserAssetAccountTypeValidator crossUserAssetAccountTypeValidator;

    public ManageLedgerImpl(
        LedgerTxnRepository ledgerTxnRepository,
        LedgerSplitRepository ledgerSplitRepository,
        AccountRepository accountRepository,
        CurrencyRepository currencyRepository,
        CategoryTagRepository categoryTagRepository,
        ManageHousehold manageHousehold,
        UserAvatarService userAvatarService,
        MinimumSplitCountValidator minimumSplitCountValidator,
        PositiveSplitAmountValidator positiveSplitAmountValidator,
        SplitSideValueValidator splitSideValueValidator,
        DoubleEntryBalanceValidator doubleEntryBalanceValidator,
        TransactionAccountCurrencyValidator transactionAccountCurrencyValidator,
        IndividualModeOwnershipValidator individualModeOwnershipValidator,
        HouseholdIdPresenceValidator householdIdPresenceValidator,
        HouseholdMembershipValidator householdMembershipValidator,
        HouseholdSharedAccountValidator householdSharedAccountValidator,
        CrossUserAssetAccountTypeValidator crossUserAssetAccountTypeValidator
    ) {
        this.ledgerTxnRepository = ledgerTxnRepository;
        this.ledgerSplitRepository = ledgerSplitRepository;
        this.accountRepository = accountRepository;
        this.currencyRepository = currencyRepository;
        this.categoryTagRepository = categoryTagRepository;
        this.manageHousehold = manageHousehold;
        this.userAvatarService = userAvatarService;
        this.minimumSplitCountValidator = minimumSplitCountValidator;
        this.positiveSplitAmountValidator = positiveSplitAmountValidator;
        this.splitSideValueValidator = splitSideValueValidator;
        this.doubleEntryBalanceValidator = doubleEntryBalanceValidator;
        this.transactionAccountCurrencyValidator = transactionAccountCurrencyValidator;
        this.individualModeOwnershipValidator = individualModeOwnershipValidator;
        this.householdIdPresenceValidator = householdIdPresenceValidator;
        this.householdMembershipValidator = householdMembershipValidator;
        this.householdSharedAccountValidator = householdSharedAccountValidator;
        this.crossUserAssetAccountTypeValidator = crossUserAssetAccountTypeValidator;
    }

    @Override
    @Transactional
    public TransactionDto createTransaction(CreateTransactionDto dto, User creator) {
        long userId = requireNotNull(creator.getUserId(), "User ID must not be null");
        boolean isHouseholdMode =
            dto.getMode() != null && "HOUSEHOLD".equals(dto.getMode().toUpperCase(Locale.ROOT));

        minimumSplitCountValidator.validate(dto.getSplits());
        positiveSplitAmountValidator.validate(dto.getSplits());
        splitSideValueValidator.validate(dto.getSplits());
        doubleEntryBalanceValidator.validate(dto.getSplits());

        var currency = currencyRepository
            .findById(dto.getCurrency())
            .orElseThrow(() -> new BadRequestException("Invalid currency: " + dto.getCurrency()));

        List<UUID> accountIds = dto.getSplits().stream().map(CreateSplitDto::getAccountId).distinct().toList();
        List<Account> accounts = accountRepository.findAllById(accountIds);
        if (accounts.size() != accountIds.size()) {
            Set<UUID> foundIds = accounts.stream().map(Account::getId).collect(Collectors.toSet());
            List<UUID> missingIds = accountIds.stream().filter(id -> !foundIds.contains(id)).toList();
            throw new BadRequestException("Accounts not found: " + missingIds);
        }

        Map<UUID, Account> accountMap = accounts
            .stream()
            .collect(Collectors.toMap(account -> requireNotNull(account.getId()), Function.identity()));

        transactionAccountCurrencyValidator.validate(accounts, dto.getCurrency());

        List<Account> nonOwnedAccounts = accounts
            .stream()
            .filter(account -> !Objects.equals(account.getOwner() != null ? account.getOwner().getUserId() : null, userId))
            .toList();

        if (!nonOwnedAccounts.isEmpty()) {
            individualModeOwnershipValidator.validate(nonOwnedAccounts, isHouseholdMode);
            UUID householdId = householdIdPresenceValidator.validate(dto.getHouseholdId());
            householdMembershipValidator.validate(manageHousehold, householdId, userId);
            householdSharedAccountValidator.validate(nonOwnedAccounts, manageHousehold, householdId);
            crossUserAssetAccountTypeValidator.validate(accounts);
        }

        List<UUID> categoryTagIds = dto
            .getSplits()
            .stream()
            .map(CreateSplitDto::getCategoryTagId)
            .filter(Objects::nonNull)
            .distinct()
            .toList();

        Map<UUID, CategoryTag> categoryTagMap;
        if (!categoryTagIds.isEmpty()) {
            List<CategoryTag> tags = categoryTagRepository.findAllById(categoryTagIds);
            if (tags.size() != categoryTagIds.size()) {
                Set<UUID> foundIds = tags.stream().map(CategoryTag::getId).collect(Collectors.toSet());
                List<UUID> missingIds = categoryTagIds.stream().filter(id -> !foundIds.contains(id)).toList();
                throw new BadRequestException("Category tags not found: " + missingIds);
            }

            for (CategoryTag tag : tags) {
                if (!Objects.equals(tag.getOwner() != null ? tag.getOwner().getUserId() : null, userId)) {
                    throw new ForbiddenException("Category tag '" + tag.getName() + "' is not owned by current user");
                }
            }

            categoryTagMap = tags
                .stream()
                .collect(Collectors.toMap(tag -> requireNotNull(tag.getId()), Function.identity()));
        } else {
            categoryTagMap = Map.of();
        }

        LedgerTxn txn = new LedgerTxn();
        txn.setCreatedBy(creator);
        txn.setHouseholdId(isHouseholdMode ? dto.getHouseholdId() : null);
        txn.setTxnDate(dto.getTxnDate());
        txn.setDescription(dto.getDescription().trim());
        txn.setCurrency(currency);

        List<LedgerSplit> splits = dto.getSplits().stream().map(splitDto -> {
            LedgerSplit split = new LedgerSplit();
            split.setTransaction(txn);
            split.setAccount(accountMap.get(splitDto.getAccountId()));
            split.setSide(SplitSide.valueOf(splitDto.getSide()));
            split.setAmountMinor(splitDto.getAmountMinor());
            split.setCategoryTag(splitDto.getCategoryTagId() != null ? categoryTagMap.get(splitDto.getCategoryTagId()) : null);
            return split;
        }).toList();

        txn.setSplits(new ArrayList<>(splits));

        LedgerTxn savedTxn = ledgerTxnRepository.save(txn);
        return toDto(savedTxn, userAvatarService);
    }

    @Override
    @Transactional(readOnly = true)
    public PagedTransactionsDto listTransactions(
        User user,
        String mode,
        UUID householdId,
        LocalDate dateFrom,
        LocalDate dateTo,
        UUID accountId,
        UUID categoryId,
        int page,
        int size
    ) {
        long userId = requireNotNull(user.getUserId(), "User ID must not be null");
        boolean isHouseholdMode = mode != null && "HOUSEHOLD".equals(mode.toUpperCase(Locale.ROOT));

        Specification<LedgerTxn> spec;
        if (isHouseholdMode) {
            UUID hhId = householdId;
            if (hhId == null) {
                throw new BadRequestException("householdId is required for household mode");
            }
            if (!manageHousehold.isActiveMember(hhId, userId)) {
                throw new ForbiddenException("Not an active member of this household");
            }
            Set<UUID> sharedAccountIds = manageHousehold.getSharedAccountIds(hhId);
            if (sharedAccountIds.isEmpty()) {
                return new PagedTransactionsDto(List.of(), page, size, 0L, 0);
            }
            spec = LedgerTxnSpecs.hasAnySharedAccount(sharedAccountIds);
        } else {
            spec = LedgerTxnSpecs.forUser(userId);
        }

        if (dateFrom != null) {
            spec = spec.and(LedgerTxnSpecs.dateFrom(dateFrom));
        }
        if (dateTo != null) {
            spec = spec.and(LedgerTxnSpecs.dateTo(dateTo));
        }
        if (accountId != null) {
            spec = spec.and(LedgerTxnSpecs.hasAccount(accountId));
        }
        if (categoryId != null) {
            spec = spec.and(LedgerTxnSpecs.hasCategory(categoryId));
        }

        PageRequest pageable = PageRequest.of(
            page,
            size,
            Sort.by(Sort.Direction.DESC, "txnDate").and(Sort.by(Sort.Direction.DESC, "createdAt"))
        );
        Page<LedgerTxn> pageResult = ledgerTxnRepository.findAll(spec, pageable);

        return new PagedTransactionsDto(
            pageResult.getContent().stream().map(txn -> toDto(txn, userAvatarService)).toList(),
            pageResult.getNumber(),
            pageResult.getSize(),
            pageResult.getTotalElements(),
            pageResult.getTotalPages()
        );
    }

    @Override
    @Transactional(readOnly = true)
    public List<BalanceDto> getAccountBalances(List<UUID> accountIds, LocalDate asOf, User user, UUID householdId) {
        if (accountIds.isEmpty()) {
            return List.of();
        }

        long userId = requireNotNull(user.getUserId(), "User ID must not be null");
        List<UUID> uniqueAccountIds = accountIds.stream().distinct().toList();
        List<Account> accounts = accountRepository.findAllById(uniqueAccountIds);
        if (accounts.size() != uniqueAccountIds.size()) {
            throw new NotFoundException("Account not found");
        }

        if (householdId != null) {
            if (!manageHousehold.isActiveMember(householdId, userId)) {
                throw new ForbiddenException("Not an active member of this household");
            }
            Set<UUID> sharedAccountIds = manageHousehold.getSharedAccountIds(householdId);
            if (uniqueAccountIds.stream().anyMatch(id -> !sharedAccountIds.contains(id))) {
                throw new ForbiddenException("Account is not shared into this household");
            }
        } else {
            boolean hasNonOwned = accounts
                .stream()
                .anyMatch(account -> !Objects.equals(account.getOwner() != null ? account.getOwner().getUserId() : null, userId));
            if (hasNonOwned) {
                throw new ForbiddenException("Not the owner of this account");
            }
        }

        Map<UUID, Long> rawBalancesByAccountId = ledgerSplitRepository
            .computeRawBalancesByAccountIds(uniqueAccountIds, asOf, SplitSide.DEBIT, SplitSide.CREDIT)
            .stream()
            .collect(Collectors.toMap(balance -> balance.getAccountId(), balance -> balance.getRawBalance()));

        Map<UUID, Account> accountById = accounts
            .stream()
            .collect(Collectors.toMap(account -> requireNotNull(account.getId()), Function.identity()));

        return uniqueAccountIds.stream().map(accountId -> {
            Account account = accountById.get(accountId);
            long rawBalance = rawBalancesByAccountId.getOrDefault(accountId, 0L);
            long balanceMinor = DEBIT_NORMAL_TYPES.contains(account.getType()) ? rawBalance : -rawBalance;

            return new BalanceDto(
                accountId,
                account.getName(),
                account.getType().name(),
                requireNotNull(account.getCurrency() != null ? account.getCurrency().getCode() : null),
                balanceMinor,
                asOf
            );
        }).toList();
    }

    @Override
    @Transactional(readOnly = true)
    public BalanceDto getAccountBalance(UUID accountId, LocalDate asOf, User user, UUID householdId) {
        long userId = requireNotNull(user.getUserId(), "User ID must not be null");

        Account account = accountRepository.findById(accountId).orElseThrow(() -> new NotFoundException("Account not found"));

        if (householdId != null) {
            if (!manageHousehold.isActiveMember(householdId, userId)) {
                throw new ForbiddenException("Not an active member of this household");
            }
            if (!manageHousehold.isAccountShared(householdId, accountId)) {
                throw new ForbiddenException("Account is not shared into this household");
            }
        } else if (!Objects.equals(account.getOwner() != null ? account.getOwner().getUserId() : null, userId)) {
            throw new ForbiddenException("Not the owner of this account");
        }

        boolean isDebitNormal = Set.of(AccountType.ASSET, AccountType.EXPENSE).contains(account.getType());
        long balanceMinor = isDebitNormal
            ? ledgerSplitRepository.computeBalance(accountId, asOf, SplitSide.DEBIT, SplitSide.CREDIT)
            : ledgerSplitRepository.computeBalance(accountId, asOf, SplitSide.CREDIT, SplitSide.DEBIT);

        return new BalanceDto(
            requireNotNull(account.getId()),
            account.getName(),
            account.getType().name(),
            requireNotNull(account.getCurrency() != null ? account.getCurrency().getCode() : null),
            balanceMinor,
            asOf
        );
    }

    private static TransactionDto toDto(LedgerTxn txn, UserAvatarService avatarService) {
        List<LedgerSplit> splits = txn.getSplits() != null ? txn.getSplits() : List.of();
        List<SplitDto> splitDtos = splits.stream().map(ManageLedgerImpl::toDto).toList();

        TxnCreatorDto createdByDto = null;
        if (txn.getCreatedBy() != null) {
            createdByDto = new TxnCreatorDto(
                txn.getCreatedBy().getFirstName(),
                txn.getCreatedBy().getLastName(),
                txn.getCreatedBy().getEmail(),
                avatarService.resolveAvatarDataUrl(txn.getCreatedBy().getAvatarPath())
            );
        }

        return new TransactionDto(
            requireNotNull(txn.getId(), "Transaction ID must not be null"),
            txn.getTxnDate(),
            requireNotNull(txn.getCurrency() != null ? txn.getCurrency().getCode() : null, "Currency must not be null"),
            txn.getDescription(),
            txn.getHouseholdId(),
            deriveTxnKind(splits),
            createdByDto,
            splitDtos,
            txn.getCreatedAt(),
            txn.getUpdatedAt()
        );
    }

    private static String deriveTxnKind(List<LedgerSplit> splits) {
        Set<AccountType> accountTypes = splits
            .stream()
            .map(split -> split.getAccount() != null ? split.getAccount().getType() : null)
            .filter(Objects::nonNull)
            .collect(Collectors.toCollection(HashSet::new));

        if (accountTypes.stream().allMatch(TRANSFER_TYPES::contains)) {
            return "TRANSFER";
        }
        if (accountTypes.stream().anyMatch(type -> type == AccountType.EXPENSE)) {
            return "EXPENSE";
        }
        if (accountTypes.stream().anyMatch(type -> type == AccountType.INCOME)) {
            return "INCOME";
        }
        return "TRANSFER";
    }

    private static SplitDto toDto(LedgerSplit split) {
        AccountType accountType = requireNotNull(
            split.getAccount() != null ? split.getAccount().getType() : null,
            "Account type must not be null"
        );
        boolean isDebitNormal = DEBIT_NORMAL_TYPES.contains(accountType);
        boolean increases =
            (isDebitNormal && split.getSide() == SplitSide.DEBIT)
                || (!isDebitNormal && split.getSide() == SplitSide.CREDIT);
        long effectMinor = increases ? split.getAmountMinor() : -split.getAmountMinor();

        return new SplitDto(
            requireNotNull(split.getId(), "Split ID must not be null"),
            requireNotNull(split.getAccount() != null ? split.getAccount().getId() : null, "Account must not be null"),
            requireNotNull(
                split.getAccount() != null ? split.getAccount().getName() : null,
                "Account name must not be null"
            ),
            accountType.name(),
            split.getSide().name(),
            split.getAmountMinor(),
            effectMinor,
            split.getCategoryTag() != null ? split.getCategoryTag().getId() : null,
            split.getCategoryTag() != null ? split.getCategoryTag().getName() : null
        );
    }

    private static <T> T requireNotNull(T value) {
        return requireNotNull(value, "Required value was null.");
    }

    private static <T> T requireNotNull(T value, String message) {
        if (value == null) {
            throw new IllegalArgumentException(message);
        }
        return value;
    }
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr_api/services/ledger/ManageLedgerImpl.kt
package com.decrux.pocketr_api.services.ledger

import com.decrux.pocketr_api.entities.db.auth.User
import com.decrux.pocketr_api.entities.db.ledger.*
import com.decrux.pocketr_api.entities.dtos.BalanceDto
import com.decrux.pocketr_api.entities.dtos.CreateTransactionDto
import com.decrux.pocketr_api.entities.dtos.SplitDto
import com.decrux.pocketr_api.entities.dtos.TransactionDto
import com.decrux.pocketr_api.repositories.*
import com.decrux.pocketr_api.services.household.ManageHousehold
import org.springframework.data.jpa.domain.Specification
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException
import java.time.LocalDate
import java.util.UUID

@Service
class ManageLedgerImpl(
    private val ledgerTxnRepository: LedgerTxnRepository,
    private val ledgerSplitRepository: LedgerSplitRepository,
    private val accountRepository: AccountRepository,
    private val currencyRepository: CurrencyRepository,
    private val categoryTagRepository: CategoryTagRepository,
    private val manageHousehold: ManageHousehold,
) : ManageLedger {

    @Transactional
    override fun createTransaction(dto: CreateTransactionDto, creator: User): TransactionDto {
        val userId = requireNotNull(creator.userId) { "User ID must not be null" }
        val isHouseholdMode = dto.mode?.uppercase() == "HOUSEHOLD"

        // 1. Require at least 2 splits
        if (dto.splits.size < 2) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Transaction must have at least 2 splits")
        }

        // 2. Validate all amounts > 0
        dto.splits.forEach { split ->
            if (split.amountMinor <= 0) {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "All split amounts must be greater than 0")
            }
        }

        // 3. Validate split sides
        dto.splits.forEach { split ->
            try {
                SplitSide.valueOf(split.side)
            } catch (_: IllegalArgumentException) {
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid split side: ${split.side}")
            }
        }

        // 4. Double-entry: sum(DEBIT) == sum(CREDIT)
        val sumDebits = dto.splits.filter { it.side == "DEBIT" }.sumOf { it.amountMinor }
        val sumCredits = dto.splits.filter { it.side == "CREDIT" }.sumOf { it.amountMinor }
        if (sumDebits != sumCredits) {
            throw ResponseStatusException(
                HttpStatus.BAD_REQUEST,
                "Double-entry violation: sum of debits ($sumDebits) must equal sum of credits ($sumCredits)",
            )
        }

        // 5. Validate currency exists
        val currency = currencyRepository.findById(dto.currency)
            .orElseThrow { ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid currency: ${dto.currency}") }

        // 6. Load and validate all accounts
        val accountIds = dto.splits.map { it.accountId }.distinct()
        val accounts = accountRepository.findAllById(accountIds)
        if (accounts.size != accountIds.size) {
            val foundIds = accounts.map { it.id }.toSet()
            val missingIds = accountIds.filter { it !in foundIds }
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Accounts not found: $missingIds")
        }
        val accountMap = accounts.associateBy { requireNotNull(it.id) }

        // 7. Currency consistency: all split accounts must match transaction currency
        accounts.forEach { account ->
            if (account.currency?.code != dto.currency) {
                throw ResponseStatusException(
                    HttpStatus.BAD_REQUEST,
                    "Account '${account.name}' has currency ${account.currency?.code} but transaction currency is ${dto.currency}",
                )
            }
        }

        // 8. Permission check: individual vs household mode
        val ownedAccounts = accounts.filter { it.owner?.userId == userId }
        val nonOwnedAccounts = accounts.filter { it.owner?.userId != userId }

        if (nonOwnedAccounts.isNotEmpty()) {
            // Cross-user transaction requires household mode
            if (!isHouseholdMode) {
                throw ResponseStatusException(
                    HttpStatus.FORBIDDEN,
                    "Cannot post to accounts not owned by current user in individual mode",
                )
            }

            val householdId = dto.householdId
                ?: throw ResponseStatusException(HttpStatus.BAD_REQUEST, "householdId is required for household mode")

            // Verify creator is active member of household
            if (!manageHousehold.isActiveMember(householdId, userId)) {
                throw ResponseStatusException(HttpStatus.FORBIDDEN, "Not an active member of this household")
            }

            // Verify all non-owned accounts are shared into the household
            nonOwnedAccounts.forEach { account ->
                val accountId = requireNotNull(account.id)
                if (!manageHousehold.isAccountShared(householdId, accountId)) {
                    throw ResponseStatusException(
                        HttpStatus.FORBIDDEN,
                        "Account '${account.name}' is not shared into household",
                    )
                }
            }

            // Cross-user transfer rule: all accounts must be ASSET only (v1)
            accounts.forEach { account ->
                if (account.type != AccountType.ASSET) {
                    throw ResponseStatusException(
                        HttpStatus.BAD_REQUEST,
                        "Cross-user transfers only allow ASSET accounts (v1), but '${account.name}' is ${account.type}",
                    )
                }
            }
        }

        // 9. Validate category tags
        val categoryTagIds = dto.splits.mapNotNull { it.categoryTagId }.distinct()
        val categoryTagMap: Map<UUID, CategoryTag> = if (categoryTagIds.isNotEmpty()) {
            val tags = categoryTagRepository.findAllById(categoryTagIds)
            if (tags.size != categoryTagIds.size) {
                val foundIds = tags.map { it.id }.toSet()
                val missingIds = categoryTagIds.filter { it !in foundIds }
                throw ResponseStatusException(HttpStatus.BAD_REQUEST, "Category tags not found: $missingIds")
            }
            tags.forEach { tag ->
                if (tag.owner?.userId != userId) {
                    throw ResponseStatusException(
                        HttpStatus.FORBIDDEN,
                        "Category tag '${tag.name}' is not owned by current user",
                    )
                }
            }
            tags.associateBy { requireNotNull(it.id) }
        } else {
            emptyMap()
        }

        // 10. Persist transaction
        val txn = LedgerTxn(
            createdBy = creator,
            householdId = if (isHouseholdMode) dto.householdId else null,
            txnDate = dto.txnDate,
            description = dto.description.trim(),
            currency = currency,
        )

        val splits = dto.splits.map { splitDto ->
            LedgerSplit(
                transaction = txn,
                account = accountMap.getValue(splitDto.accountId),
                side = SplitSide.valueOf(splitDto.side),
                amountMinor = splitDto.amountMinor,
                categoryTag = splitDto.categoryTagId?.let { categoryTagMap[it] },
                memo = splitDto.memo?.trim()?.ifBlank { null },
            )
        }
        txn.splits = splits.toMutableList()

        val savedTxn = ledgerTxnRepository.save(txn)
        return savedTxn.toDto()
    }

    @Transactional(readOnly = true)
    override fun listTransactions(
        user: User,
        mode: String?,
        householdId: UUID?,
        dateFrom: LocalDate?,
        dateTo: LocalDate?,
        accountId: UUID?,
        categoryId: UUID?,
    ): List<TransactionDto> {
        val userId = requireNotNull(user.userId) { "User ID must not be null" }
        val isHouseholdMode = mode?.uppercase() == "HOUSEHOLD"

        var spec: Specification<LedgerTxn> = if (isHouseholdMode) {
            val hhId = householdId
                ?: throw ResponseStatusException(HttpStatus.BAD_REQUEST, "householdId is required for household mode")
            if (!manageHousehold.isActiveMember(hhId, userId)) {
                throw ResponseStatusException(HttpStatus.FORBIDDEN, "Not an active member of this household")
            }
            LedgerTxnSpecs.forHousehold(hhId)
        } else {
            LedgerTxnSpecs.forUser(userId)
        }

        dateFrom?.let { spec = spec.and(LedgerTxnSpecs.dateFrom(it)) }
        dateTo?.let { spec = spec.and(LedgerTxnSpecs.dateTo(it)) }
        accountId?.let { spec = spec.and(LedgerTxnSpecs.hasAccount(it)) }
        categoryId?.let { spec = spec.and(LedgerTxnSpecs.hasCategory(it)) }

        return ledgerTxnRepository.findAll(spec).map { it.toDto() }
    }

    @Transactional(readOnly = true)
    override fun getAccountBalance(accountId: UUID, asOf: LocalDate, user: User, householdId: UUID?): BalanceDto {
        val userId = requireNotNull(user.userId) { "User ID must not be null" }

        val account = accountRepository.findById(accountId)
            .orElseThrow { ResponseStatusException(HttpStatus.NOT_FOUND, "Account not found") }

        // In household mode, allow viewing shared accounts; otherwise require ownership
        if (householdId != null) {
            if (!manageHousehold.isActiveMember(householdId, userId)) {
                throw ResponseStatusException(HttpStatus.FORBIDDEN, "Not an active member of this household")
            }
            if (!manageHousehold.isAccountShared(householdId, accountId)) {
                throw ResponseStatusException(HttpStatus.FORBIDDEN, "Account is not shared into this household")
            }
        } else if (account.owner?.userId != userId) {
            throw ResponseStatusException(HttpStatus.FORBIDDEN, "Not the owner of this account")
        }

        val isDebitNormal = account.type in setOf(AccountType.ASSET, AccountType.EXPENSE)
        val balanceMinor = if (isDebitNormal) {
            ledgerSplitRepository.computeBalance(accountId, asOf, SplitSide.DEBIT, SplitSide.CREDIT)
        } else {
            ledgerSplitRepository.computeBalance(accountId, asOf, SplitSide.CREDIT, SplitSide.DEBIT)
        }

        return BalanceDto(
            accountId = requireNotNull(account.id),
            accountName = account.name,
            accountType = account.type.name,
            currency = requireNotNull(account.currency?.code),
            balanceMinor = balanceMinor,
            asOf = asOf,
        )
    }

    private companion object {
        val DEBIT_NORMAL_TYPES = setOf(AccountType.ASSET, AccountType.EXPENSE)
        val TRANSFER_TYPES = setOf(AccountType.ASSET, AccountType.LIABILITY, AccountType.EQUITY)

        fun LedgerTxn.toDto(): TransactionDto {
            val splitDtos = splits.map { it.toDto() }
            return TransactionDto(
                id = requireNotNull(id) { "Transaction ID must not be null" },
                txnDate = txnDate,
                currency = requireNotNull(currency?.code) { "Currency must not be null" },
                description = description,
                householdId = householdId,
                txnKind = deriveTxnKind(splits),
                splits = splitDtos,
                createdAt = createdAt,
                updatedAt = updatedAt,
            )
        }

        fun deriveTxnKind(splits: List<LedgerSplit>): String {
            val accountTypes = splits.mapNotNull { it.account?.type }.toSet()
            return when {
                accountTypes.all { it in TRANSFER_TYPES } -> "TRANSFER"
                accountTypes.any { it == AccountType.EXPENSE } -> "EXPENSE"
                accountTypes.any { it == AccountType.INCOME } -> "INCOME"
                else -> "TRANSFER"
            }
        }

        fun LedgerSplit.toDto(): SplitDto {
            val accountType = requireNotNull(account?.type) { "Account type must not be null" }
            val isDebitNormal = accountType in DEBIT_NORMAL_TYPES
            val increases = (isDebitNormal && side == SplitSide.DEBIT) || (!isDebitNormal && side == SplitSide.CREDIT)
            val effectMinor = if (increases) amountMinor else -amountMinor

            return SplitDto(
                id = requireNotNull(id) { "Split ID must not be null" },
                accountId = requireNotNull(account?.id) { "Account must not be null" },
                accountName = requireNotNull(account?.name) { "Account name must not be null" },
                accountType = accountType.name,
                side = side.name,
                amountMinor = amountMinor,
                effectMinor = effectMinor,
                categoryTagId = categoryTag?.id,
                categoryTagName = categoryTag?.name,
                memo = memo,
            )
        }
    }
}
